src/compiler/checker.ts

```diff
@@ -9098,7 +9098,7 @@ namespace ts {
                         }
                     }
                     const sourceTypes = some(constructorTypes, t => !!(t.flags & ~TypeFlags.Nullable)) ? constructorTypes : types; // TODO: GH#18217
-                    type = getUnionType(sourceTypes!, UnionReduction.Subtype);
+                    type = getUnionType(sourceTypes!);

```

Comment:
If you have a JS file like:

```js
if (process.env.NODE_ENV === "production") {
  module.exports = require("./foo.min.js");
} else {
  module.exports = require("./foo.js");
}
```

If one module is a subtype of the other, go-to-definition on an import/require of this file was previously going straight into the supertype module. By not eagerly doing subtype reduction here, the services are able to retain the information that the export branches off into these two separate modules. I traced this line of code all the way back to the beginning of JS support, long before `UnionReduction.Literal` existed. At the time it was introduced, `getUnionType` did subtype reduction by default. Later, that default was flipped, and this call site opted into it as a mechanical migration of the default. Even later, the default became `UnionReduction.Literal`, but this call site was of course ignored. No pre-existing tests were impacted by my changing it now, so my assumption is this does not actually have an intentional reason to do something other than the default.